src/utils/JSONReviver.ts

```diff
@@ -87,16 +88,23 @@ export function Generic_toJSON<T extends Record<string, any>>(
  * @returns    The object */
 export function Generic_fromJSON<T extends Record<string, any>>(
   ctor: new () => T,
-  // data can actually be anything. We're just pretending it has the right keys for T. Save data is not type validated.
-  data: Record<keyof T, any>,
+  data: unknown,
   keys?: readonly (keyof T)[],
 ): T {
+  objectAssert(data);
   const obj = new ctor();
   // If keys were provided, just load the provided keys (if they are in the data)
   if (keys) {
-    for (const key of keys) {
+    /**
+     * The type of key is "keyof T", but the type of data is Record<string, unknown>. TypeScript won't allow us to use
+     * key as the index of data, so we need to typecast here.
+     */
+    for (const key of keys as string[]) {
       const val = data[key];
-      if (val !== undefined) obj[key] = val;
+      if (val !== undefined) {
+        // @ts-expect-error -- TypeScript won't allow this action: Type 'T' is generic and can only be indexed for reading.
```

Comment:
Can you eliminate this by deferring the cast to `const val = data[key as string]`? Or something similar? 
Basically, this used to work, it seems like the issue comes from the type of key being string.

Reply:
I tried to mix and match many ways of typecasting, but TS always complained about something. In the past, data is `Record<keyof T, any>`, so we don't have any problems with key (`keyof T`). Now we change data to `unknown`. After asserting, data is `Record<string, unknown>`. I tried `const val = data[key as string]`, but TS complained:
```
Type '{} | null' is not assignable to type 'T[keyof T]'.
  '{} | null' is assignable to the constraint of type 'T[keyof T]', but 'T[keyof T]' could be instantiated with a different subtype of constraint 'any'.
    Type 'null' is not assignable to type 'T[keyof T]'.
      'null' is assignable to the constraint of type 'T[keyof T]', but 'T[keyof T]' could be instantiated with a different subtype of constraint 'any'.
```
If I change the check from `val !== undefined` to `val != null`, TS will complain:
```
Type '{}' is not assignable to type 'T[keyof T]'.
  '{}' is assignable to the constraint of type 'T[keyof T]', but 'T[keyof T]' could be instantiated with a different subtype of constraint 'any'.
```

Reply:
I think the loop below provides the missing piece: a cast of `val` to `T[keyof T]`.

We should probably also now that the cast (or TS-disable) is *not* safe, in the sense that we can load wrongly-typed data here at runtime. I don't see a way around that without some massive changes.